openupgrade_scripts/scripts/mail/16.0.1.10/upgrade_analysis_work.txt

```diff
@@ -0,0 +1,223 @@
+mail         / mail.channel.partner     / custom_channel_name (char)    : DEL
+mail         / mail.channel.partner     / fetched_message_id (many2one) : DEL relation: mail.message
+mail         / mail.channel.partner     / fold_state (selection)        : DEL selection_keys: ['closed', 'folded', 'open']
+mail         / mail.channel.partner     / guest_id (many2one)           : DEL relation: mail.guest
+mail         / mail.channel.partner     / is_minimized (boolean)        : DEL
+mail         / mail.channel.partner     / is_pinned (boolean)           : DEL
+mail         / mail.channel.partner     / last_interest_dt (datetime)   : DEL
+mail         / mail.channel.partner     / partner_id (many2one)         : DEL relation: res.partner
+mail         / mail.channel.partner     / rtc_inviting_session_id (many2one): DEL relation: mail.channel.rtc.session
+mail         / mail.channel.partner     / rtc_session_ids (one2many)    : DEL relation: mail.channel.rtc.session
+mail         / mail.channel.partner     / seen_message_id (many2one)    : DEL relation: mail.message
+# DONE (pre-migration): mail.channel.partner renamed to mail.channel.member.
+# New datetime field last_seen_dt: NOTHING TO DO
+
+mail         / mail.channel.rtc.session / channel_member_id (many2one)  : NEW relation: mail.channel.member, required
+mail         / mail.channel.rtc.session / channel_partner_id (many2one) : DEL relation: mail.channel.partner, required
+# DONE (pre-migration): copy column channel_partner_id into channel_member_id
+
+mail         / mail.gateway.allowed     / email (char)                  : NEW
+mail         / mail.gateway.allowed     / email_normalized (char)       : NEW isfunction: function, stored
+# NOTHING TO DO: new model
+
+mail         / mail.guest               / channel_ids (many2many)       : table is now 'mail_channel_member' ('mail_channel_partner')
+# DONE in pre-migrate script when renaming the model
+
+mail         / mail.link.preview        / image_mimetype (char)         : NEW
+mail         / mail.link.preview        / message_id (many2one)         : NEW relation: mail.message, required
+mail         / mail.link.preview        / og_description (text)         : NEW
+mail         / mail.link.preview        / og_image (char)               : NEW
+mail         / mail.link.preview        / og_mimetype (char)            : NEW
+mail         / mail.link.preview        / og_title (char)               : NEW
+mail         / mail.link.preview        / og_type (char)                : NEW
+mail         / mail.link.preview        / source_url (char)             : NEW required
+# NOTHING TO DO: new model
+
+mail         / mail.mail                / fetchmail_server_id (many2one): previously in module fetchmail
+# NOTHING TO DO
+mail         / mail.mail                / scheduled_date (char)         : type is now 'datetime' ('char')
+# DONE (pre- and post-migration): Keep old char value in DB (with legacy column name) and convert the value to datetime
```

Comment:
There is a function that change date to datetime. 
``use date_to_datetime_tz``  
https://oca.github.io/OpenUpgrade/API.html#openupgradelib.openupgrade.date_to_datetime_tz

Here, we could make char -> date then use ``date_to_datetime_tz`` ? 
I think code could be more simple. (and maybe faster, because the "for each" by ORM, in post migration can be slow on table with a lot of row in mail.mail.

Don't you think ?

Reply:
Good idea, I agree to (try to) doing it by SQL. The char version of `scheduled_date` contains a datetime, not a date. So I would rather use TO_TIMESTAMP to directly convert it? But then idk how  to deal with the timezone: the timezone might be specified in the char field, or it is assumed to be UTC. But TO_TIMESTAMP sql method doesn't allow to deal with timezones?

Reply:
...you can set a string scheduled date for v15 in CI in a file similar to `data.py` in https://github.com/OCA/OpenUpgrade/tree/16.0/openupgrade_scripts/scripts/base/16.0.1.3/tests, and then assert something about it in the test file

Reply:
@hbrunn you're right, doing nothing works. Going to add a test

Reply:
Done, could you review it please?

Reply:
thanks, looks good. I did however just try this out with a db that has `scheduled_date==''` and that breaks, so maybe you could create demo data to this effect and add an update statement to set empty strings to null?

Reply:
I added an update statement for empty strings and added a test case